bftengine/tests/testViewChange/testViewChange.cpp

```diff
@@ -165,73 +138,182 @@ TEST(testViewchangeSafetyLogic_test, computeRestrictions) {
+      primary, curView, assignedSeqNum, bftEngine::impl::CommitPath::SLOW, false, clientRequest->size());
+  ppMsg->addRequest(clientRequest->body(), clientRequest->size());
+  ppMsg->finishAddingRequests();
+
+  char buff[32]{};
+  PrepareFullMsg* pfMsg = PrepareFullMsg::create(curView, assignedSeqNum, primary, buff, sizeof(buff));
 
   ViewChangeMsg** viewChangeMsgs = new ViewChangeMsg*[N];
-  for (int i = 0; i < N; i++) {
-    if (i == 1) {
-      viewChangeMsgs[i] = nullptr;
+
+  viewChangeMsgs[0] = new ViewChangeMsg(0, curView, lastStableSeqNum);
+  viewChangeMsgs[1] = nullptr;
+  viewChangeMsgs[2] = new ViewChangeMsg(2, curView, lastStableSeqNum);
+  viewChangeMsgs[3] = new ViewChangeMsg(3, curView, lastStableSeqNum);
+
+  viewChangeMsgs[2]->addElement(*pRepInfo,
+                                assignedSeqNum,
+                                ppMsg->digestOfRequests(),
+                                ppMsg->viewNumber(),
+                                true,
+                                ppMsg->viewNumber(),
+                                pfMsg->signatureLen(),
+                                pfMsg->signatureBody());
+
+  auto VCS = ViewChangeSafetyLogic(N, F, C, &dummyVerifier_, PrePrepareMsg::digestOfNullPrePrepareMsg());
+
+  SeqNum min{}, max{};
+  VCS.computeRestrictions(viewChangeMsgs, 0, min, max, restrictions);
+
+  for (int i = 0; i < kWorkWindowSize; i++) {
+    if (i == assignedSeqNum - 1) {
+      Assert(!restrictions[i].isNull);
+      Assert(ppMsg->digestOfRequests().toString() == restrictions[i].digest.toString());
     } else {
-      viewChangeMsgs[i] = new ViewChangeMsg(i, 1, lastStableSeqNum);
+      Assert(restrictions[i].isNull);
     }
   }
 
-  viewChangeMsgs[2]->addElement(
-      *pRepInfo, assignedSeqNum, pp->digestOfRequests(), 0, true, 0, pfMsg->signatureLen(), pfMsg->signatureBody());
+  for (int i = 0; i < N; i++) {
+    delete viewChangeMsgs[i];
+    viewChangeMsgs[i] = nullptr;
+  }
+  delete[] viewChangeMsgs;
+  delete clientRequest;
+  delete pfMsg;
+  delete ppMsg;
+}
+// The purpose of this test is to verify that when a Replica receives a set of
+// view change messages enough for it to confirm a view change it will consider
+// the Safe Prepare Certificate for a sequence number that has two Prepare
+// Certificates. The safe Prepare Certificate for a given sequence is the one with
+// the highest View number, so after ViewchangeSafetyLogic::computeRestrictions
+// is called we verify that for this sequence number we are going to redo the
+// protocol with the right PrePrepare, matching the Prepare Certificate with
+// the highest View number.
+TEST(testViewchangeSafetyLogic_test, computeRestrictions_two_prepare_certs_for_same_seq_no) {
+  char buff[32]{};
+  bftEngine::impl::SeqNum lastStableSeqNum = 100;
+  const uint32_t kRequestLength = 2;
+
+  uint64_t expectedLastValue1 = 12345;
+  const uint64_t requestBuffer1[kRequestLength] = {(uint64_t)200, expectedLastValue1};
+  ViewNum curView1 = 0;
```

Comment:
It's not the current view per se, it's "a view" (possibly not the current). So I'd suggest to call this view1.